sql/ GenrePreference.sql

```diff
@@ -0,0 +1,17 @@
+SELECT CASE
+           WHEN DATEDIFF(YEAR, Borrowers.DateOfBirth, GETDATE()) BETWEEN 0 AND 10 THEN '0-10'
+           WHEN DATEDIFF(YEAR, Borrowers.DateOfBirth, GETDATE()) BETWEEN 11 AND 20 THEN '11-20'
+           WHEN DATEDIFF(YEAR, Borrowers.DateOfBirth, GETDATE()) BETWEEN 21 AND 30 THEN '21-30'
+           ELSE '31+'
+           END AS AgeGroup, Books.Genre, COUNT(*) AS BorrowCount
+FROM Loans
+         JOIN Borrowers ON Loans.BorrowerID = Borrowers.BorrowerID
+         JOIN Books ON Loans.BookID = Books.BookID
+GROUP BY Books.Genre,
+         CASE
+             WHEN DATEDIFF(YEAR, Borrowers.DateOfBirth, GETDATE()) BETWEEN 0 AND 10 THEN '0-10'
+             WHEN DATEDIFF(YEAR, Borrowers.DateOfBirth, GETDATE()) BETWEEN 11 AND 20 THEN '11-20'
+             WHEN DATEDIFF(YEAR, Borrowers.DateOfBirth, GETDATE()) BETWEEN 21 AND 30 THEN '21-30'
```

Comment:
think of a way to avoid repeating your self

Reply:
You are right we can avoid repeating the CASE statement by using a common table expression (CTE) 

```suggestion
WITH BorrowerAgeGroups AS (
    SELECT 
        CASE
            WHEN DATEDIFF(YEAR, Borrowers.DateOfBirth, GETDATE()) BETWEEN 0 AND 10 THEN '0-10'
            WHEN DATEDIFF(YEAR, Borrowers.DateOfBirth, GETDATE()) BETWEEN 11 AND 20 THEN '11-20'
            WHEN DATEDIFF(YEAR, Borrowers.DateOfBirth, GETDATE()) BETWEEN 21 AND 30 THEN '21-30'
            ELSE '31+'
        END AS AgeGroup,
        Books.Genre
    FROM Loans
    JOIN Borrowers ON Loans.BorrowerID = Borrowers.BorrowerID
    JOIN Books ON Loans.BookID = Books.BookID
)
SELECT 
    AgeGroup, 
    Genre, 
    COUNT(*) AS BorrowCount
FROM BorrowerAgeGroups
GROUP BY AgeGroup, Genre
HAVING COUNT(*) > 0;

```